File: src/actionengine/nodes/async_node.cc
// Copyright 2026 The Action Engine Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "async_node.h"

#include <memory>
#include <optional>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

#include <absl/strings/str_cat.h>

#include "actionengine/data/types.h"
#include "actionengine/net/stream.h"
#include "actionengine/nodes/node_map.h"
#include "actionengine/stores/chunk_store.h"
#include "actionengine/stores/chunk_store_reader.h"
#include "actionengine/stores/chunk_store_writer.h"
#include "actionengine/stores/local_chunk_store.h"
#include "actionengine/util/status_macros.h"

namespace act {

AsyncNode::AsyncNode(std::string_view id, NodeMap* absl_nullable node_map,
                     std::unique_ptr<ChunkStore> chunk_store)
    : node_map_(node_map), chunk_store_(std::move(chunk_store)) {
  if (chunk_store_ == nullptr) {
    chunk_store_ = std::make_unique<LocalChunkStore>();
  }
  // TODO: figure out a way to propagate errors from constructors.
  chunk_store_->SetId(id).IgnoreError();
}

AsyncNode::~AsyncNode() {
  act::MutexLock lock(&mu_);
  if (default_reader_ != nullptr) {
    default_reader_->Cancel();
    mu_.unlock();
    default_reader_.reset();
    mu_.lock();
  }
  if (default_writer_ != nullptr) {
    mu_.unlock();
    default_writer_->WaitForBufferToDrain();
    mu_.lock();
    default_writer_.reset();
  }
}

absl::Status AsyncNode::PutInternal(NodeFragment fragment)
    ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
  const std::string node_id(chunk_store_->GetId());
  if (!fragment.id.empty()) {
    if (fragment.id != node_id) {
      return absl::FailedPreconditionError(
          absl::StrCat("Fragment id: ", fragment.id,
                       " does not match the node id: ", chunk_store_->GetId()));
    }
    if (node_id.empty()) {
      RETURN_IF_ERROR(chunk_store_->SetId(fragment.id));
    }
  }

  ASSIGN_OR_RETURN(Chunk & chunk, fragment.GetChunk());
  return PutInternal(std::move(chunk), fragment.seq.value_or(-1),
                     !fragment.continued);
}

absl::Status AsyncNode::PutInternal(Chunk chunk, int seq, bool final)
    ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
  ChunkStoreWriter* writer = EnsureWriter();

  RETURN_IF_ERROR(writer->Put(std::move(chunk), seq, final).status());

  return absl::OkStatus();
}

ChunkStoreWriter& AsyncNode::GetWriter() ABSL_LOCKS_EXCLUDED(mu_) {
  act::MutexLock lock(&mu_);
  return *EnsureWriter();
}

absl::Status AsyncNode::GetWriterStatus() const {
  act::MutexLock lock(&mu_);
  if (default_writer_ == nullptr) {
    return absl::OkStatus();
  }
  return default_writer_->GetStatus();
}

auto AsyncNode::GetId() const -> std::string {
  return std::string(chunk_store_->GetId());
}

absl::StatusOr<std::optional<Chunk>> AsyncNode::Next(
    std::optional<absl::Duration> timeout) {
  ChunkStoreReader& reader = GetReader();
  return reader.Next(timeout);
}

ChunkStoreReader& AsyncNode::GetReader() ABSL_LOCKS_EXCLUDED(mu_) {
  act::MutexLock lock(&mu_);
  return *EnsureReader();
}

absl::Status AsyncNode::GetReaderStatus() const {
  act::MutexLock lock(&mu_);
  if (default_reader_ == nullptr) {
    return absl::FailedPreconditionError("Reader is not initialized.");
  }
  return default_reader_->GetStatus();
}

std::unique_ptr<ChunkStoreReader> AsyncNode::MakeReader(
    ChunkStoreReaderOptions options) const {
  return std::make_unique<ChunkStoreReader>(chunk_store_.get(), options);
}

AsyncNode& AsyncNode::SetReaderOptions(const ChunkStoreReaderOptions& options) {

  act::MutexLock lock(&mu_);
  EnsureReader()->SetOptions(options);
  return *this;
}

AsyncNode& AsyncNode::ResetReader() {
  act::MutexLock lock(&mu_);
  default_reader_ = nullptr;
  return *this;
}

ChunkStoreReader* AsyncNode::EnsureReader() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
  if (default_reader_ == nullptr) {
    default_reader_ = std::make_unique<ChunkStoreReader>(chunk_store_.get());
  }
  return default_reader_.get();
}

ChunkStoreWriter* AsyncNode::EnsureWriter(int n_chunks_to_buffer)
    ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
  if (default_writer_ == nullptr) {
    default_writer_ = std::make_unique<ChunkStoreWriter>(chunk_store_.get(),
                                                         n_chunks_to_buffer);
  }
  return default_writer_.get();
}

auto AsyncNode::Put(Chunk value, int seq, bool final) -> absl::Status {
  act::MutexLock lock(&mu_);
  const bool continued = !final && !value.IsNull();
  return PutInternal(NodeFragment{
      .id = std::string(chunk_store_->GetId()),
      .data = std::move(value),
      .seq = seq,
      .continued = continued,
  });
}

auto AsyncNode::Put(NodeFragment value) -> absl::Status {
  act::MutexLock lock(&mu_);
  return PutInternal(std::move(value));
}

template <>
AsyncNode& operator>>(AsyncNode& node, std::optional<Chunk>& value) {
  auto next_chunk = node.Next<Chunk>();
  CHECK_OK(next_chunk.status()) << "Failed to get next chunk.";
  value = *std::move(next_chunk);
  return node;
}

template <>
AsyncNode& operator<<(AsyncNode& node, NodeFragment value) {
  node.Put(std::move(value)).IgnoreError();
  return node;
}

template <>
AsyncNode& operator<<(AsyncNode& node, Chunk value) {
  const bool final = value.IsNull();
  CHECK_OK(node.Put(std::move(value), /*seq=*/-1, /*final=*/final));
  return node;
}

}  // namespace act

File: py/actionengine/async_node.py
# Copyright 2026 The Action Engine Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""A Pythonic wrapper for the raw pybind11 AsyncNode bindings."""

import asyncio
import contextlib
from collections.abc import Awaitable
from typing import Any, Optional

from actionengine import _C
from actionengine import global_settings
from actionengine import data
from actionengine import utils

Chunk = data.Chunk
ChunkMetadata = data.ChunkMetadata
NodeFragment = data.NodeFragment
LocalChunkStore = _C.chunk_store.LocalChunkStore

global_setting_if_none = global_settings.global_setting_if_none


class AsyncNode(_C.nodes.AsyncNode):
    """A Pythonic wrapper for the raw pybind11 AsyncNode bindings.

    AsyncNode is an accessor class that allows to access the chunks of a node
    asynchronously, namely, to read and write chunks from/to the underlying
    chunk store.
    """

    def __init__(
        self,
        node_id: str,
        chunk_store: Optional[_C.chunk_store.ChunkStore] = None,
        node_map: "Optional[_C.nodes.NodeMap]" = None,
        serializer_registry: Optional[data.SerializerRegistry] = None,
    ):
        """Constructor for AsyncNode.

        Makes a new AsyncNode with the given id, referencing the given chunk store
        and node map. If chunk store is not provided, a local chunk store is used.
        If node map is not provided, the node will not be able to reference other
        nodes (including its children), but will otherwise function normally.

        The node, however, will NOT be added to the node map. You will need to do
        that in an outer scope.

        Args:
          node_id: The id of the node.
          chunk_store: The chunk store to use for the node.
          node_map: The node map to use for the node.
          serializer_registry: The serializer registry to use for the node.
        """
        self._deserialize_automatically_preference: bool | None = None
        self._serializer_registry = (
            serializer_registry or data.get_global_serializer_registry()
        )
        self._reader_options_set = False
        super().__init__(node_id, node_map, chunk_store)

    def _add_python_specific_attributes(self):
        """Adds Python-specific attributes to the node."""
        self._deserialize_automatically_preference: bool | None = None
        self._serializer_registry = data.get_global_serializer_registry()
        self._reader_options_set = False

    @property
    def deserialize(self) -> bool:
        """Returns whether the node deserializes objects automatically."""
        return global_setting_if_none(
            self._deserialize_automatically_preference,
            "readers_deserialise_automatically",
        )

    def _consume_sync(
        self, timeout: float = -1.0, allow_none: bool = False
    ) -> Any:
        item = self.next_sync(timeout)
        if item is None:
            if allow_none:
                return None
            raise RuntimeError(
                "Node is empty while expecting exactly one item."
            )
        if self.next_sync(timeout) is not None:
            raise RuntimeError(
                "Node has more than one item while expecting exactly one."
            )
        return item

    def consume(
        self, timeout: float = -1.0, allow_none: bool = False
    ) -> Any | Awaitable[Any]:
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return self._consume_sync(timeout, allow_none)
        return asyncio.create_task(
            asyncio.to_thread(self._consume_sync, timeout, allow_none)
        )

    async def next(self, timeout: float = -1.0):
        if self.deserialize:
            return await self.next_object(timeout)
        else:
            return await self.next_chunk(timeout)

    def next_sync(self, timeout: float = -1.0):
        if self.deserialize:
            return self.next_object_sync(timeout)
        else:
            return self.next_chunk_sync(timeout)

    async def next_object(self, timeout: float = -1.0) -> Any:
        """Returns the next object in the store, or None if the store is empty."""
        return await asyncio.create_task(
            asyncio.to_thread(self.next_object_sync, timeout)
        )

    def next_object_sync(self, timeout: float = -1.0) -> Any:
        """Returns the next object in the store, or None if the store is empty."""
        chunk = self.next_chunk_sync(timeout)
        if chunk is None:
            return None
        return data.from_chunk(
            chunk,
            mimetype=chunk.metadata.mimetype,
            registry=self._serializer_registry,
        )

    async def next_chunk(self, timeout: float = -1.0) -> Optional[Chunk]:
        return await asyncio.create_task(
            asyncio.to_thread(self.next_chunk_sync, timeout)
        )

    def next_chunk_sync(self, timeout: float = -1.0) -> Optional[Chunk]:
        return _C.nodes.AsyncNode.next_chunk(
            self, timeout
        )  # pytype: disable=attribute-error

    async def next_fragment(
        self, timeout: float = -1.0
    ) -> Optional[NodeFragment]:
        """Returns the next fragment in the store, or None if the store is empty."""
        return await asyncio.to_thread(self.next_fragment_sync, timeout)

    def next_fragment_sync(
        self, timeout: float = -1.0
    ) -> Optional[NodeFragment]:
        """Returns the next fragment in the store, or None if the store is empty."""
        return _C.nodes.AsyncNode.next_fragment(
            self, timeout
        )  # pytype: disable=attribute-error

    def put_fragment(
        self, fragment: NodeFragment, seq: int = -1
    ):  # pylint: disable=useless-parent-delegation
        """Puts a fragment into the node's chunk store.

        This method will only block if the node's chunk store writer's buffer is
        full. Otherwise, it will return immediately.

        Args:
          fragment: The fragment to put.
          seq: The sequence id of the fragment.

        Returns:
          None if the fragment was put synchronously with no event loop, or an
          awaitable if the fragment was put asynchronously within an event loop.
        """
        return _C.nodes.AsyncNode.put_fragment(
            self, fragment, seq
        )  # pytype: disable=attribute-error

    def put_chunk(
        self, chunk: Chunk, seq: int = -1, final: bool = False
    ):  # pylint: disable=useless-parent-delegation
        """Puts a chunk into the node's chunk store.

        This method will only block if the node's chunk store writer's buffer is
        full. Otherwise, it will return immediately.

        Args:
          chunk: The chunk to put.
          seq: The sequence id of the chunk.
          final: Whether the chunk is final.

        Returns:
          None if the chunk was put synchronously with no event loop, or an
          awaitable if the chunk was put asynchronously within an event loop.
        """
        return _C.nodes.AsyncNode.put_chunk(
            self, chunk, seq, final
        )  # pytype: disable=attribute-error,

    def put_and_finalize(
        self,
        obj: Any,
        seq: int = -1,
        mimetype: str | None = None,
    ) -> None | Awaitable[None]:
        return self.put(obj, seq, True, mimetype)

    def put_sync(
        self,
        obj: Any,
        seq: int = -1,
        final: bool = False,
        mimetype: str | None = None,
    ) -> None:
        """Puts an object into the node's chunk store."""

        if isinstance(obj, Chunk):
            if mimetype is not None:
                raise ValueError(
                    "mimetype must not be specified when putting a Chunk object."
                )
            return self.put_chunk(obj, seq, final)

        if isinstance(obj, NodeFragment):
            if mimetype is not None:
                raise ValueError(
                    "mimetype must not be specified when putting a NodeFragment object."
                )
            return self.put_fragment(obj, seq)

        chunk = data.to_chunk(
            obj,
            mimetype=mimetype or "",
            registry=self._serializer_registry,
        )
        return self.put_chunk(chunk, seq, final)

    def put(
        self,
        obj: Any,
        seq: int = -1,
        final: bool = False,
        mimetype: str | None = None,
    ) -> None | Awaitable[None]:
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return self.put_sync(obj, seq, final, mimetype)

        return asyncio.to_thread(
            self.put_sync,
            obj,
            seq,
            final,
            mimetype,
        )

    async def copy_from(self, src: "AsyncNode"):
        """Populates the node with the contents of another node."""

        with src.deserialize_automatically(False) as src:
            async for chunk in src:
                self.put_chunk(chunk)
            await self.finalize()

    def put_text(self, text: str, seq: int = -1, final: bool = False):
        """Puts a text/plain chunk into the node's chunk store."""
        return self.put_chunk(
            Chunk(
                metadata=ChunkMetadata(mimetype="text/plain"),
                data=text.encode("utf-8"),
            ),
            seq=seq,
            final=final,
        )

    def finalize_sync(self) -> None:
        """Finalizes the node's writer stream."""
        return self.put_chunk(
            Chunk(
                metadata=ChunkMetadata(mimetype="application/octet-stream"),
                data=b"",
            ),
            final=True,
        )

    def finalize(self) -> None | Awaitable[None]:
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return self.finalize_sync()
        return asyncio.to_thread(self.finalize_sync)

    # pylint: disable-next=[useless-parent-delegation]
    def get_id(self) -> str:
        """Returns the id of the node."""
        return _C.nodes.AsyncNode.get_id(
            self
        )  # pytype: disable=attribute-error

    def set_reader_options(
        self,
        ordered: bool | None = None,
        remove_chunks: bool | None = None,
        n_chunks_to_buffer: int | None = None,
        timeout: float | None = None,
        start_seq_or_offset: int = 0,
    ) -> "AsyncNode":
        """Sets the options for the default reader on the node.

        Args:
          ordered: Whether to read chunks in order.
          remove_chunks: Whether to remove chunks from the store after reading them.
          n_chunks_to_buffer: The number of chunks to buffer until blocking the
            internal ChunkStoreReader.
          timeout: The timeout for reading chunks, in seconds. If None, the default
            timeout is used, which is -1.0 (no timeout).
          start_seq_or_offset: The sequence id or offset to start reading from.

        Returns:
          The node itself.
        """

        _C.nodes.AsyncNode.set_reader_options(  # pytype: disable=attribute-error
            self,
            ordered,
            remove_chunks,
            n_chunks_to_buffer,
            timeout,
            start_seq_or_offset,
        )
        return self

    @contextlib.contextmanager
    def write_context(self):
        """Returns the node which automatically finalizes the writer stream."""
        try:
            yield self
        finally:
            self.finalize()

    @contextlib.contextmanager
    def deserialize_automatically(self, deserialize: bool = True):
        """Returns the node which automatically deserialises objects."""
        self._deserialize_automatically_preference = deserialize
        try:
            yield self
        finally:
            self._deserialize_automatically_preference = None

    def __aiter__(self):
        return self

    def __iter__(self):
        return self

    async def __anext__(self):
        try:
            chunk = await self.next_chunk()
            while chunk is not None and utils.is_null_chunk(chunk):
                chunk = await self.next_chunk()
        except asyncio.CancelledError as exc:
            print("AsyncNode.__anext__: CancelledError")
            raise StopAsyncIteration() from exc

        if chunk is None:
            raise StopAsyncIteration()

        if self.deserialize:
            return await asyncio.to_thread(
                data.from_chunk,
                chunk,
                "",
                self._serializer_registry,
            )

        return chunk

    def __next__(self):
        chunk = self.next_chunk_sync()
        while chunk is not None and utils.is_null_chunk(chunk):
            chunk = self.next_chunk_sync()

        if chunk is None:
            raise StopIteration()

        if self.deserialize:
            return data.from_chunk(
                chunk,
                mimetype=chunk.metadata.mimetype,
                registry=self._serializer_registry,
            )

        return chunk
